File: app/backend/src/services/argocd.ts
import axios, {AxiosInstance} from 'axios';
import https from 'https';
import {config} from '../config';
import {logger} from './logger';

let client: AxiosInstance | null = null;

function getClient(): AxiosInstance | null {
    if (!config.argocd.token) return null;
    if (!client) {
        client = axios.create({
            baseURL: `${config.argocd.url}/api/v1`,
            headers: {Authorization: `Bearer ${config.argocd.token}`},
            httpsAgent: config.argocd.insecure
                ? new https.Agent({rejectUnauthorized: false})
                : undefined,
            timeout: 10_000,
        });
    }
    return client;
}

export async function getArgoHealthMap(): Promise<Map<string, { health: string; sync: string }>> {
    const map = new Map<string, { health: string; sync: string }>();
    const api = getClient();
    if (!api) return map;

    try {
        const {data} = await api.get('/applications');
        for (const app of data.items ?? []) {
            map.set(app.metadata.name, {
                health: app.status?.health?.status ?? 'Unknown',
                sync: app.status?.sync?.status ?? 'Unknown',
            });
        }
    } catch (err) {
        logger.error('ArgoCD fetch failed', {error: (err as Error).message});
    }

    return map;
}

File: app/backend/internal/services/argocd.go
package services

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/mephalrith/noodles/backend/internal/config"
)

type ArgoStatus struct {
	Health string
	Sync   string
}

type ArgoCDService struct {
	cfg    *config.Config
	client *http.Client
}

func NewArgoCDService(cfg *config.Config) *ArgoCDService {
	transport := http.DefaultTransport.(*http.Transport).Clone()
	if cfg.ArgoCD.Insecure {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}

	return &ArgoCDService{
		cfg: cfg,
		client: &http.Client{
			Transport: transport,
			Timeout:   10 * time.Second,
		},
	}
}

func (a *ArgoCDService) GetHealthMap() map[string]ArgoStatus {
	result := make(map[string]ArgoStatus)

	if a.cfg.ArgoCD.Token == "" {
		return result
	}

	url := fmt.Sprintf("%s/api/v1/applications", a.cfg.ArgoCD.URL)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		Logger.Error("ArgoCD: failed to create request", "error", err)
		return result
	}
	req.Header.Set("Authorization", "Bearer "+a.cfg.ArgoCD.Token)

	resp, err := a.client.Do(req)
	if err != nil {
		Logger.Error("ArgoCD fetch failed", "error", err)
		return result
	}
	defer resp.Body.Close()

	var data struct {
		Items []struct {
			Metadata struct {
				Name string `json:"name"`
			} `json:"metadata"`
			Status struct {
				Health struct {
					Status string `json:"status"`
				} `json:"health"`
				Sync struct {
					Status string `json:"status"`
				} `json:"sync"`
			} `json:"status"`
		} `json:"items"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		Logger.Error("ArgoCD: failed to decode response", "error", err)
		return result
	}

	for _, app := range data.Items {
		health := app.Status.Health.Status
		if health == "" {
			health = "Unknown"
		}
		sync := app.Status.Sync.Status
		if sync == "" {
			sync = "Unknown"
		}
		result[app.Metadata.Name] = ArgoStatus{Health: health, Sync: sync}
	}

	return result
}
